Limiter extension API interfaces (**draft 4**)

jmacd opened the pull request:
#### Description

See #12603. This follows discussion in #12700 and has been updated extensively based on feedback.

#### Link to tracking issue

Part of #9591.

#### Testing

TODO

#### Documentation

Done.

bogdandrutu commented on extension/extensionlimiter/README.md:
```diff
@@ -0,0 +1,305 @@
+        metadata.Type,
+        createDefaultConfig,
+        xreceiver.WithMetrics(createMetrics, metadata.MetricsStability),
+        xreceiver.WithLimiters(getLimiters),
```
Should the limiters be part of the receiver instance? A factory should be agnostic of the config, and per "`getLimiters` is a function to get the effective`[]configmiddleware.Config`" this comes from a config.

jmacd replied:
I am not sure, this is a nuanced question because some receivers like OTLP support different middleware settings for each protocol.

If we are going to configure request_items and memory_size via receivers, while configuring request_count and network_bytes in middleware, then I see two choices for where the limiters are configured:

1. All limiters are in a middlewares list (consistently; even if all you are using are limiters, you can use middlewares, i.e., even for non-grpc and non-http you could use middlewares just as limiters). This way users just set middlewares. If receivers have different protocols, they use different limiters. (I prefer this option.)
2. Receivers have a separate config section, similar to the various exporterhelper configs which is embedded in receivers. Then users have to understand where limiters are configured, which is IMO more complicated. Some limiters are in the receiver-level config, some limiters are in the middleware-level config. (I prefer this other option.)

I'll do what the group would prefer, of course!

bogdandrutu commented on extension/extensionlimiter/rate.go:
```diff
@@ -0,0 +1,95 @@
+//
+// See the README for more recommendations.
+type RateLimiter interface {
+	// Limiter includes MustDeny().
+	Limiter
+
```
here is my confusion, I think the `RateLimiterProvider` should implement the Limiter, because I want to call the "saturation" only once asap. I know I can still do that by checking all the weight type, but for a limiter that supports only items I don't think you expect to call this in the "middleware".

jmacd replied:
`RateLimiterProvider` is called in Start() typically (for static limiter configurations). Are we on the same page?

Re: only once: I created an adapter for combining multiple limiters (which may be called at different times) by their MustDeny(). That's how I call MustDeny once, it's the `MultiLimiter` in limiterhelper/consumer.go. When a wrapper limiter constructs >1 limiter in the wrapper, there is a return value limiter which combines all of them for MustDeny calls.

jmacd replied:
Update: To combine `Checker` instances into a single instance, use `limiterhelper.MultiChecker`. The three basic Checker constructors are NewLimiterWrapperChecker, NewRateLimiterChecker, and NewResourceLimiterChecker.

jmacd replied:
I think I sorted this out. Thank you for pointing to this. I have modified the RateLimiter and ResourceLimiter interfaces to include GetChecker() methods which allow obtaining a Checker, having the MustDeny method. The MustDeny method and Checker interface apply to all weight keys. I have documented that MustDeny may be called more than once per request since middleware, receiver, and wrappers can all reasonably call this in various circumstances.

axw commented on extension/extensionlimiter/README.md:
```diff
@@ -0,0 +1,305 @@
+risks wasting memory. In general, an overloaded limiter that is
+saturated SHOULD fail requests immediately.
+
+Limiter implementations SHOULD consider the context deadline when
```
Should SHOULD be MUST? I tend to think considering context deadline should be mandatory. But maybe I'm being pedantic - did you just mean here that limiters MAY choose not to block if they can anticipate that the context deadline will come before the limiter is no longer saturated?

jmacd replied:
Following @bogdandrutu's suggestion, I've separated the MustDeny method (which takes only context) from the Acquire/Limit methods which are weight-dependent.

axw commented on extension/extensionlimiter/README.md:
```diff
@@ -0,0 +1,305 @@
+limit request is defined as the equivalent of passing a zero value to
+the limiter.
+
+Limiters SHOULD treat a request for zero units of the limit as a
```
Would it be reasonable to make this a MUST, and get rid of the MustDeny method?

jmacd replied:
Maybe. I like the idea, but we can keep the Limiter interface. Limiters can be defined as calling the underlying limiter with 0, and frequently we have several of them and want to call them all at once. I will separate the high-level MustDeny interface from the two low-level limiter interfaces, leaving the MultiLimiter helper, I think it will be nice.

jmacd replied:
I think in general that the suggestion by @bogdandrutu was a good one, so I went that direction.

axw commented on extension/extensionlimiter/README.md:
```diff
@@ -0,0 +1,305 @@
+
+### Limiter blocking and failing
+
+Limiters implementations MAY block the request or fail immediately,
+subject to internal logic. A limiter aims to avoid waste, which
```
What is the internal logic? Are you saying this is a property of the limiter, rather than the caller?

If we made it caller-defined, then I think we could get rid of the MustDeny call: instead you would make a non-blocking request for 0 items. If the limiter is saturated, that would return an error; otherwise it would return success without affecting the capacity.

jmacd replied:
I have raised a question about non-blocking methods in the open-questions section of the README.

jmacd commented on extension/extensionlimiter/README.md:
```diff
@@ -0,0 +1,305 @@
+
+### Limiter saturation
+
+All limiters feature a `MustDeny` method which is made available for
+applications to test when a limit is fully saturated. This special
+limit request is defined as the equivalent of passing a zero value to
+the limiter.
```
I don't see any strong reasons not to use a standalone API. The user will have to handle separate objects anyway, in most cases, since they call limiters for individual weights but ask MustDeny for all of the un-checked limits in general, I think.

I will change this to use a standalone API.

jmacd replied:
I separated the `Checker` interface which contains the `MustDeny` method. Wasn't sure what else to name it, `Limiter` feels confusing, so does `MustDenier`.

jmacd commented on extension/extensionlimiter/rate.go:
```diff
@@ -0,0 +1,95 @@
+	Limit(ctx context.Context, value uint64) error
+}
+
+// RateLimiterFunc is an easy way to construct RateLimiters.
+type RateLimiterFunc func(ctx context.Context, value uint64) error
+
+var _ RateLimiter = RateLimiterFunc(nil)
+
+// MustDeny implements RateLimiter.
+func (f RateLimiterFunc) MustDeny(ctx context.Context) error {
+	return f.Limit(ctx, 0)
+}
+
+// Limit implements RateLimiter.
+func (f RateLimiterFunc) Limit(ctx context.Context, value uint64) error {
+	if f == nil {
+		return nil
+	}
+	return f(ctx, value)
+}
```
I see what you mean. Would you prefer to have MustDeny be a completely separate interface, and limiters have two functions? I can go that way. Initially I had a single interface, just `Acquire(x)` where x==0 can be interpreted as MustDeny. I think that's what @axw was asking for [here](https://github.com/open-telemetry/opentelemetry-collector/pull/12953#discussion_r2071054918).

Re: http.HandlerFunc.

I like the pattern! No complaints. I wouldn't call net/http an extensive-user of this pattern, however. In the middleware work, `extensionmiddlewaretest` has its own copy of [what we might call `http.RoundTripperFunc` (which does not exist)](https://github.com/open-telemetry/opentelemetry-collector/blob/727ae96d621450fc5afd3a81b4669b48756a760d/extension/extensionmiddleware/extensionmiddlewaretest/nop.go#L20).

bogdandrutu replied:
>I think that's what @axw was asking for https://github.com/open-telemetry/opentelemetry-collector/pull/12953#discussion_r2071054918.

I understand that, but I go back to my point which is `MustDeny` is an interface that can be called as early as possible and is independent of the `WeightKey`. Your proposal and @axw supported version requires me to call Acquire for 4 different keys.

bogdandrutu replied:
Separate topic:

>I see what you mean. Would you prefer to have MustDeny be a completely separate interface, and limiters have two functions? I can go that way. Initially I had a single interface, just Acquire(x) where x==0 can be interpreted as MustDeny.

I don't think I am trying to suggest to have `just Acquire(x) where x==0`. I am suggesting a helper func to implement only a func in the interface, but interface still have 2 functions MustDeny and Acquire.

jmacd replied:
I see. I have misunderstood.

jmacd replied:
I have greatly misunderstood this detail. I will need to do more work on this, thank you for this comment.

jmacd commented on extension/extensionlimiter/wrapper.go:
```diff
@@ -0,0 +1,97 @@
+// providers. Usually configmiddleware or limiterhelper is responsible
+// for constructing the correct wrapper from these two kinds of limiter;
+// users will use this interface consistently.
+type LimiterWrapper interface {
```
Sure, will do.

jmacd replied:
Done.

jmacd commented on extension/extensionlimiter/rate.go:
```diff
@@ -0,0 +1,95 @@
+// middleware (e.g., grpc.StatsHandler).
+//
+// See the README for more recommendations.
+type RateLimiter interface {
```
I do not support this idea.

The the meaning behind these limits is different. I liken this to the distinction between OTel metrics UpDownCounter and Counter. The admission controller in contrib/internal/otelarrow wouldn't do anything if you called it in a gRPC statshandler for network bytes, just be meaningless overhead, because it thinks that the release function means something (i.e., it only supports resource limits) and the gRPC stats handler has no way to call the release func after the network bytes are no longer used (i.e., it only supports rate limits).

jmacd commented:
Reviewers, please see the open questions:

https://github.com/open-telemetry/opentelemetry-collector/pull/12953/files#diff-50c94a038eae5ba5747b0b2c502b0753f2af3664671c4aaec46d4916b828dc25

bogdandrutu commented on extension/extensionlimiter/extensionlimiter.go:
```diff
@@ -0,0 +1,60 @@
+
+// Checker is for checking when a limit is saturated.  This can be
+// called prior to the start of work to check for limiter saturation.
+type Checker interface {
```
Up for discussion (not sure): Should this be the "basic" limiter, then call it "Limiter" instead?

jmacd replied:
I had actually named this `Limiter` in an earlier draft. Yes.

bogdandrutu commented on extension/extensionlimiter/limiterhelper/checker.go:
```diff
@@ -0,0 +1,33 @@
+
+// MustDeny implements Checker.
+func (ls MultiChecker) MustDeny(ctx context.Context) error {
+	var err error
+	for _, lim := range ls {
+		if lim == nil {
+			continue
+		}
+		err = errors.Join(err, lim.MustDeny(ctx))
+	}
+	return err
```
`errors.Join` -> is a bit inefficient when `lim.MustDeny(ctx)` returns nil. I prefer the `multierr.Append` which does a better job.

jmacd replied:
:+1:

bogdandrutu commented on extension/extensionlimiter/extensionlimiter.go:
```diff
@@ -0,0 +1,60 @@
+	return f(ctx)
+}
+
+// CheckerProvider is an interface to obtain checkers for a group of
+// weight keys.
+type CheckerProvider interface {
+	// GetChecker returns a checker for a group of weight keys.
+	GetChecker(...Option) (Checker, error)
+}
```
What is the point of having the extra layer of "Provider" for the "Checker"?

jmacd replied:
I was thinking of the Option as a placeholder for the open questions, maybe the way to pass the Signal and Component identity. If the actual limiter instance has the options, I expect the basic limiter instance to have the options too.

bogdandrutu commented on extension/extensionlimiter/weight.go:
```diff
@@ -0,0 +1,82 @@
+// checked at a certain stage.  The receiver and middleware can both
+// be responsible for applying limits, and this type helps ensure
+// limits are applied only across cooperating sub-components.
+type WeightSet []WeightKey
```
Can you please add this later when the first usage comes? I don't want to argue more about this PR, and I am unable to understand where this is used.

jmacd replied:
This is very much not a big deal to me, we can use `[]WeightKey` instead.

The first usage is inside this PR, the helper method Contains used to prevent double-limiting request count in middlware.

Examine how the receiver and middleware cooperate to not limit the same thing twice. The receiver knows which weight keys are applied in middleware (because it knows) and it configures a limiterhelper wrapper for the remaining weight keys. I've listed the standard weight keys here, to try and help users see the mechanic.

The wrapper and the middleware are both able to configure request_count limits. For a push-based receiver, we do this in middleware, but for an arbitrary receiver we might put this logic in the wrapper because it is natural there too: in that case, you would pass three keys to the wrapper). The wrapper could be used to add a made-up network bytes limit too, maybe using the uncompressed size of the data as a proxy for receivers that do not use middleware.

bogdandrutu commented on extension/extensionlimiter/weight.go:
```diff
@@ -0,0 +1,82 @@
+// StandardNotMiddlewareKeys methods return the list of middleware
+// keys that can be automatically configured through middleware and
+// not.
+type WeightKey string
```
Not sure how important this is, do you see a big usage of this on the critical path? If yes, should we use an enum as int instead, since that will be a bit faster for type checks?

jmacd replied:
My expectation is that weight keys are used during Start() where the providers are called to initialize the actual limiter instances. We have discussed some ideas about new things that could be rate limited, and I believe @axw previously suggested this might be an open set, not a closed one; either way the goal should be to bind limiters once via a provider and use them w/o passing around weight keys at runtime.

bogdandrutu commented on extension/extensionlimiter/rate.go:
```diff
@@ -0,0 +1,72 @@
+	CheckerProvider
+
+	// GetRateLimiter returns a rate limiter for a weight key.
+	GetRateLimiter(WeightKey, ...Option) (RateLimiter, error)
```
Do you think we should have different `Option`s for different providers (different types so we can expand them independently)?

jmacd replied:
I was thinking about passing in Signal kind, Component ID via these options. I wouldn't mind making these two parameters obligatory, but I don't know how to make it so--the confighttp and configgrpc wrappers don't have this info either.

bogdandrutu commented on extension/extensionlimiter/limiterhelper/wrapper.go:
```diff
@@ -0,0 +1,125 @@
+// configmiddleware or limiterhelper is responsible for constructing
+// the correct wrapper from these two kinds of limiter; users will use
+// this interface consistently.
+type LimiterWrapper interface {
```
Do you expect others to implement this interface? Otherwise we should have this as a struct.

jmacd replied:
I see the point in the style you're after -- instead of mocking an interface, you should pass in an adapter and place a mock call inside your function.

bogdandrutu commented on extension/extensionlimiter/limiterhelper/wrapper.go:
```diff
@@ -0,0 +1,125 @@
+// the appropriate interface for callers that can easily wrap a
+// function call, because for wrapped calls there is no distinction
+// between rate limiters and resource limiters.
+type LimiterWrapperProvider interface {
```
Do you expect others to implement this interface? Otherwise we should have this as a struct.

jmacd replied:
I see. I do not need this to be an interface, however I wonder about all the mock-based tests I've ever written with `gomock` and so on, whether we lose access to this style of test. I'll make this a struct.

bogdandrutu reviewed (approved):
The public API looks good, some naming suggestions and some go implementation in the helper.

bogdandrutu commented on extension/extensionlimiter/README.md:
```diff
@@ -0,0 +1,352 @@
+An `Option` type has been added as a placeholder in the provider
+interfaces. **NOTE: No options are implemented.** Potential options:
+
+- The protocol name
+- The signal kind
+- The caller's component ID
```
I am not convinced about protocol, but I see the other 2 being useful.

jmacd replied:
Sure. Note that protocol is already distinguished through middleware configuration, so a user could configure separate limiters if the distinction matters.

bogdandrutu commented on extension/extensionlimiter/README.md:
```diff
@@ -0,0 +1,352 @@
+functions. No examples are provided. How will limiters configure, for
+example, tenant-specific limits?
+
+##### Data-dependent limits
```
If [#39199](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/39199) is accepted, do you still need support here?

bogdandrutu commented on extension/extensionlimiter/extensionlimiter.go:
```diff
@@ -0,0 +1,60 @@
+
+// Checker is for checking when a limit is saturated.  This can be
+// called prior to the start of work to check for limiter saturation.
+type Checker interface {
```
Do we need to support a limiter that is just "Checker" like the current memory limiter? I would suggest yes.

jmacd replied:
Yes. So, would you say there are three fundamental limiter APIs, the Rate/Resource and Basic? There is a provider for each -- does it make sense, to you, that there Rate/Resource providers embed the basic limiter provider (as I have called `CheckerProvider` in this PR)?

Callers would be expected to check for more-specific extension APIs before the less-specific one.

bogdandrutu replied:
I like it

axw reviewed (approved):
Apologies for the delay. LGTM overall, just a few minor comments - I think naming could be improved, but that won't affect the API significantly.

jmacd commented:
@axw Apologies, I have created a new draft github.com/open-telemetry/opentelemetry-collector/pull/13051. I mean to incorporate all of the feedback above. In particular, I would note that I've reverted the name of the basic limiter to `BaseLimiter` in the new PR and kept the original name `MustDeny`. I appreciate your request to make this more consistent, and I've added consistency in the other interfaces (e.g., ResourceLimiter has a method named ReserveResource, RateLimiter has ReserveRate).
